### PR TITLE
[2] Add fit_impl

### PR DESCRIPTION
**Pull request by jonpsy**
#5090 

@vigsterkr btw I added your std::atomic<bool> suggestion ;)

@gf712  @karlnapf

### CODE REVIEW

**Comment by jonpsy** [comment]
- m_fitted is now more structured.
- For classes which don't need fit method, throws not_implemented error

**Comment by jonpsy** [comment]
Sorry to say, I'd have to revert the std::atomic<bool> it's causing errors which are out of the scope of PR.

File: src/shogun/preprocessor/RFFPreprocessor.h
Comment (vigsterkr) [inline src/shogun/preprocessor/RFFPreprocessor.h]: remove `virtual`

File: src/shogun/preprocessor/RescaleFeatures.h
Comment (vigsterkr) [inline src/shogun/preprocessor/RescaleFeatures.h]: remove `virtual`

**Review by vigsterkr — CHANGES_REQUESTED** [review]
some minor things.

although i've just realised that Transformer::m_fitted is a simple bool.... it should be eventual changed to `std::atomic<bool>` imo

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (gf712) [inline src/shogun/preprocessor/DensePreprocessor.cpp]:
```suggestion
void DensePreprocessor<ST>::fit_impl(const SGMatrix<ST>& feature_matrix)
```

Comment (gf712) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: Shouldn't this be pure virtual and then make this is a noop, rather  `not_implemented` when it is implemented? The issue with throwing is that you might add this to a `Pipeline`, and this will throw an error because the `Pipeline` object will always do fit + transform.

**Comment by gf712** [comment]
> Sorry to say, I'd have to revert the std::atomic it's causing errors which are out of the scope of PR.

What are the issues with making m_fitted atomic?

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (jonpsy) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: Purely virtual isn't an option because few preproc don't employ fit_impl method. This would cause compile error, I guess we could make it empty function then?

Comment (gf712) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: yes, I think so

**Comment by jonpsy** [comment]
Once this PR #5097  is ready, I could make this support std::atomic<bool>

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (karlnapf) [inline src/shogun/preprocessor/DensePreprocessor.cpp]: Nit: "No features provided" (better style)

File: src/shogun/preprocessor/DensePreprocessor.h
Comment (karlnapf) [inline src/shogun/preprocessor/DensePreprocessor.h:82]: no error in here?

Comment (karlnapf) [reply src/shogun/preprocessor/DensePreprocessor.h:82]: I like the SGMarix!

Comment (karlnapf) [reply src/shogun/preprocessor/DensePreprocessor.h:82]: what about inplace? it is const

File: src/shogun/preprocessor/PruneVarSubMean.cpp
Comment (karlnapf) [inline src/shogun/preprocessor/PruneVarSubMean.cpp:33]: nice!

**Review by karlnapf — COMMENTED** [review]
Nice! Some minori-sh comments/questions

File: src/shogun/preprocessor/DensePreprocessor.h
Comment (gf712) [reply src/shogun/preprocessor/DensePreprocessor.h:82]: The idea is that you shouldn’t get an error if the Transformer doesn’t implement a fit method because it doesn’t make sense. But I think this should be pure virtual function and then each implementation handles this.. that way nothing unexpected happens..

Comment (karlnapf) [reply src/shogun/preprocessor/DensePreprocessor.h:82]: could you explain why it doesnt make sense? 
I am ok with a purely virtual method, just needs more refactoring

Any thoughts about the const matrix and inplace?

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (jonpsy) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: > Shouldn't this be pure virtual and then make this is a noop, rather `not_implemented` when it is implemented? The issue with throwing is that you might add this to a `Pipeline`, and this will throw an error because the `Pipeline` object will always do fit + transform.

@karlnapf  this is why it doesn't make sense(from the second sentence).

Comment (jonpsy) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: @gf712 , what I did instead was make the fit_impl a noop in ```DensePreprocessor```. This way we needn't make it noop in each and every subclass which doesn't use fit_impl, and those who use fit_impl simply override this :)

File: src/shogun/preprocessor/DensePreprocessor.h
Comment (gf712) [reply src/shogun/preprocessor/DensePreprocessor.h:82]: > could you explain why it doesnt make sense?

ah sorry, yes. It doesn't make sense for something like `LogPlusOne`. There is nothing to `fit`, only to `apply`. So if I put this `Transformer` in a `Pipeline` it will always throw (if we have an exception, instead of a noop)

> Any thoughts about the const matrix and inplace?

Isn't the inplace for `transform` rather than `fit`?

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (gf712) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: sure, I think that should be fine :)

Comment (karlnapf) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: how many subclasses dont have a `fit` method? I guess that would be very few (`LogPlusOne`, which else?)
I'd actually prefer to be as explicit as possible with this. I.e. making the method purely virtual in the base class, and then the classes that don't need it override with as a no-op. That seems to make the structure much more clear. I.e. it is the exception that a subclass has a noop fit method, not the common case. If we make it a noop in the base class we are putting the cart in front of the horse imo

File: src/shogun/preprocessor/DensePreprocessor.h
Comment (karlnapf) [reply src/shogun/preprocessor/DensePreprocessor.h:82]: ah yes, sorry I keep confusing the two.
See above comment on error/purely-virtual/no-op

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (jonpsy) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: A total of five(Pnorm, NormOne, LogPlusOne, Homogenous, SumOne) don't have fit method. 

I see your point, I drew my inspiration from the ```fit``` implementation we have on ```Transformer.h``` it is defined as noop, so those who don't use it get a noop others override it. We can go with your method as well, but I'd like to know why we didn't use this same methodology on ```fit```?

**Comment by jonpsy** [comment]
std::atomic integrated :)

File: src/shogun/preprocessor/DensePreprocessor.h
Comment (LiuYuHui) [inline src/shogun/preprocessor/DensePreprocessor.h]:
```suggestion
		void fit(std::shared_ptr<Features>) override;
```

Comment (jonpsy) [reply src/shogun/preprocessor/DensePreprocessor.h]: Nice catch ;)

File: src/shogun/preprocessor/DensePreprocessor.cpp
Comment (karlnapf) [reply src/shogun/preprocessor/DensePreprocessor.cpp]: Interesting, thx for pointing that out. In that case let's stick with what is there then for consistency reasons (i.e. with what you suggested)

Asking why something is the way it is can be difficult in shogun ... many authors, many opinions ;)

**Comment by jonpsy** [comment]
GTM??

**Comment by gf712** [comment]
> GTM??

I think so, I'll let @karlnapf merge it just in case he still needs to do a final check :)

**Review by karlnapf — APPROVED** [review]
LGTM

**Comment by karlnapf** [comment]
I'll wait for the CI and then hit the button :)
Thanks!

**Pull request merged**
